gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystem.java

```diff
@@ -231,21 +233,34 @@ public WritableByteChannel create(URI path, CreateFileOptions options) throws IO
               "Cannot create a file whose name looks like a directory: '%s'", resourceId));
     }
 
-    // Check if a directory of that name exists.
-    if (options.isEnsureNoDirectoryConflict()
-        && getFileInfoInternal(resourceId.toDirectoryId(), /* inferImplicitDirectories */ true)
-            .exists()) {
-      throw new FileAlreadyExistsException("A directory with that name exists: " + path);
+    // Before creating a top-level directory we need to check if there are no conflicting files
+    // with the same name as any subdirectory
+    if (options.isEnsureNoConflictingItems()) {
+      // Asynchronously check if a directory with the same name exists.
+      StorageResourceId dirId = resourceId.toDirectoryId();
+      Future<Boolean> conflictingDirExist =
```

Comment:
Since we already use Executor here, is it a better practice to throw both of two tasks into threadpool and let them run?

Reply:
We use executor only when we need to parallelize some requests - running requests asynchronously in executor and immediately blocking on its execution should be not faster than just executing directly.